hyperspy/io_plugins/nexus.py

```diff
@@ -0,0 +1,1251 @@
+                    if item.attrs["NX_class"] != b"NXdata":
+                        tree[new_key]=find_meta_in_tree(item,rootkey,lazy=lazy)                                     
+                else:
+                    tree[new_key]=find_meta_in_tree(item,rootkey,lazy=lazy)                                     
+                    
+        return tree   
+    extracted_tree = find_meta_in_tree(group,rootname,lazy=lazy)
+    if search_keys in ["all"]:
+        return extracted_tree
+    else:
+        return _find_search_keys_in_dict(extracted_tree,search_keys=search_keys)    
+
+
+
+
+def _fix_exclusion_keys(key):    
+    """Exclude hyperspy specific keys
+    Signal and DictionaryBrowser break if a
+    a key is a dict method - e.g. {"keys":2.0} 
+    
+    This method prepends the key with "fix_" so the information is
+    still present to work around this issue
+    
+    Parameter
+    ---------
+    key : str 
+    
+    Returns
+    -------
+    str
+        key prepended with 'fix_' if key is 
+        in a set of excluded of keywords
+    
+    """
+    exclusion_list=["keys"]
+    
+    if any([key.startswith(x) for x in exclusion_list]):
+        return "fix_"+key
+    else:
+        return key
+
+
+def _find_search_keys_in_dict(tree,search_keys="all"):    
+    """Search through a dict and return a dictionary
+    whose full path contains one of the search keys
+
+    This is a convenience method to inspect a file for a value
+    rather than loading the file as a signal
+    
+    Parameters
+    ----------
+    tree         : h5py File object 
+    search_keys  : string or list of strings
+        Only return items which contain the strings
+        .e.g search_keys = ["instrument","Fe"] will return
+        hdf entries with instrument or Fe in their hdf path.
+    
+    Returns
+    -------
+    dict
+        When search_list is specified only full paths
+        containing one or more search_keys will be returned 
+        
+    """
+
+    _check_search_keys(search_keys)
+    metadata_dict = {}
+    rootname=""
+
+    # recursive function    
+    def find_searchkeys_in_tree(myDict,rootname):
+        for key, value in myDict.items():
+            if rootname != "":
+                rootkey = rootname +"/"+ key
+            else:
+                rootkey = key
+            if type(search_keys) is list \
+            and any([s1 in rootkey for s1 in search_keys]):
+                mod_keys = _text_split(rootkey, (".","/") )                        
+                # create the key, values in the dict
+                p = metadata_dict
+                for d in mod_keys[:-1]:
+                    p = p.setdefault(d,{})
+                p[mod_keys[-1]] = value
+            if isinstance(value,dict):
+                find_searchkeys_in_tree(value,rootkey)
+                    
+    if search_keys in ["all","hardlinks"]:
+        return tree
+    else:
+        find_searchkeys_in_tree(tree,rootname)
+    return metadata_dict
```

Comment:
Can it simplified by using the h5py [`visit`](https://h5py.readthedocs.io/en/latest/high/group.html#h5py.Group.visit) function with a nexus specific callable?

Reply:
The issue with visit is that it doesn't follow soft or external links within the file and many of our use cases for Nexus have linked files.
See  https://github.com/h5py/h5py/issues/671

Reply:
OK, it would be good to add the h5py github issue as reference!

Reply:
I've added to the docstring or did you want it in the biblio?